src/components/timer/TimerController.cpp

```diff
@@ -0,0 +1,59 @@
+//
+// Created by florian on 16.05.21.
+//
+
+#include "TimerController.h"
+#include "systemtask/SystemTask.h"
+#include "app_timer.h"
+#include "task.h"
+
+using namespace Pinetime::Controllers;
+
+
+APP_TIMER_DEF(timerAppTimer);
+
+
+TimerController::TimerController(System::SystemTask& systemTask) : systemTask{systemTask} {
+}
+
+
+void TimerController::Init() {
+  app_timer_create(&timerAppTimer, APP_TIMER_MODE_SINGLE_SHOT, timerEnd);
+  
+}
+
+void TimerController::StartTimer(uint32_t duration) {
+  app_timer_stop(timerAppTimer);
+  app_timer_start(timerAppTimer, APP_TIMER_TICKS(duration), this);
+  //dont ask why this random scaling factor is here. i got a difference between the time set and the time displayed and it works out to
+  //be exactly this linear factor
+  //you might be wondering why im not simply using app_timer_cnt_get() here. I am too. It is in app_timer.h, but the compiler says it
+  // doesnt exist
+  endTime = ((static_cast<float >(xTaskGetTickCount()) / static_cast<float >(configTICK_RATE_HZ)) * 1000 + APP_TIMER_TICKS(duration)) -
```

Comment:
Because one second is 1024 not 1000 ticks AFAIK

Reply:
I mostly just stole that line from your StopWatch code. Should i change the 1000 to 1024?

Reply:
Oh whoops, sorry, I was confused there for a sec, `configTICK_RATE_HZ` makes it correct.

Reply:
That's another reason why I would rather like we use custom implementation instead of blindly use these "apps" from the SDK : we use it as a black box without understanding what it does under the hood.

To me, it is needed we understand what this code does, and answer all the questions from this comment. We need to ensure we do the right thing and understand why we do it.

At first glance, app_timer uses RTC1. RTC1 is configured by FreeRTOS as its systick timer (I just hope app_timer does not alter this configuration): 1024 ticks of RTC1 correspond to 1s. That's probably why you need to apply these 1000/1024/0.024 factors in the code.

Reply:
the funny business here is that i have to use both freeRTOS and "apps" since I cant get the ticks from the apps timer for reasons that are beyond me. In theory  `APP_TIMER_TICKS` should give me milliseconds and I trust `(static_cast<float >(xTaskGetTickCount()) / static_cast<float >(configTICK_RATE_HZ)) * 1000)` to do the same since its already in StopWatch. Obviously one of those assumptions has to be wrong. If it helps you sleep, I can rip out the app_timer and replace it with xTimer from FreeRTOS. Maybe also open up a PR to change it in the Motor driver while I'm at it. I'm not sure I can still do that today, but there's really no need to rush anything here.

Reply:
We can keep app_timer in a first iteration, I just want us to ensure that the code does what we think it's doing :) Let me have a closer look at that part...

Reply:
There are 2 implementations of app_timer in the NRF SDK : app_timer.c and app_timer_freertos.c. As Infinitime uses FreeRTOS, you need to use the _freertos implementation (you'll get link error if you try to add app_timer.c in the project because both freertos and app_timer defines RTC1_Handler()). Sadly, this implementation does not provide `app_timer_cnt_get` and `app_timer_cnt_diff_compute` which would have been useful. 

`app_timer_cnt_get()` can be replaced by `xTaskGetTickCount()` as app_timer and FreeRTOS both use RTC1 as base timer.

I couldn't find an easy replacement for `app_timer_cnt_diff_compute()` but we'll need something similar the handle the counter overflow : RTC1 is a counter that is incremented 1024x per second. Once it reaches its max value (I  *think* this is a 24bit timer), it'll rollback to 0 and start counting again. We probably want to take this edge case into account.

So, RTC1 gives 1024 ticks per 1000ms, that's why we have to convert to/from ticks/ms. In order to simplify the code, I would recommend to use only 'ticks' in the internal of `TimerController` and convert them to milliseconds only in  `GetTimeRemaining()`.

Maybe you can have a look at these changes? 

```
void TimerController::StartTimer(uint32_t duration) {
  app_timer_stop(timerAppTimer);
  auto currentTicks = xTaskGetTickCount();
  app_timer_start(timerAppTimer, APP_TIMER_TICKS(duration), this);
  endTicks = currentTicks + APP_TIMER_TICKS(duration);
  timerRunning = true;
}

uint32_t TimerController::GetTimeRemaining() {
  if (!timerRunning) {
    return 0;
  }
  auto currentTicks = xTaskGetTickCount();
  auto deltaTicks = endTicks - currentTicks; // TODO Handle the counter overflow
  return (static_cast<float>(deltaTicks) / static_cast<float>(configTICK_RATE_HZ)) * 1000.0f;
}
```
I tried with a 1 minute timer and it looks like to counts and displays the time correctly.

What do you think?

Reply:
This looks a lot more like the function in [StopWatch.ccp](https://github.com/JF002/InfiniTime/blob/develop/src/displayapp/screens/StopWatch.cpp) I originally lifted the conversion from now. Here it is 
```cpp
TickType_t calculateDelta(const TickType_t startTime, const TickType_t currentTime) {
    TickType_t delta = 0;
    // Take care of overflow
    if (startTime > currentTime) {
      delta = 0xffffffff - startTime;
      delta += (currentTime + 1);
    } else {
      delta = currentTime - startTime;
    }
    return delta;
  }
  ```
Let me quickly explain why I made the changes I made and why I was wrong.
- since I app_timer and xTaskGetTickCount() come from different libraries, I didn’t expect the underlying ticks to be the same. So I converted everything to milliseconds. Since they both use the same hardware timer configured in the same way, that assumption was wrong and things can be simplified by just using ticks.
- I thought I wouldn’t have to handle overflows since I’m only counting down. This obviously doesn’t take the potential for overflows from calculating the end time into account. 

So overall this seems like a decent way to do it. Just two more things:
Do we have to use floats here? I’d much rather not. The error we get from integer division is on the scale of microseconds which shouldn’t matter at all here. 
Secondly, this thing about the FreeRTOS implementation only implementing a subset of the API seems like an easy thing to trip over. Are there other instances of this kind of thing and is there anything we can do to make the inspector aware of it?

Reply:
The fact that both timer_app and freertos use the same underlying hardware is a happy coincidence (well... I guess NRF designed it that way in purpose, but you cannot know it without looking at the implementation. Your first idea to convert to a common unit between both libs doesn make sense!
Using strong types here (Ticks, Milliseconds or std::chrono::millisecond) would make things a lo more clear and explicit, by the way.

Regarding floats, you are right, they are probably not needed here! and I also think it's better to avoid them when possible!

> Secondly, this thing about the FreeRTOS implementation only implementing a subset of the API seems like an easy thing to trip over. Are there other instances of this kind of thing and is there anything we can do to make the inspector aware of it?

NRF SDK provide both implementation, one that use the NRF scheduler and the other that is based on freertos. I don't know why they designed it that way. It looks like functions like `app_timer_cnt_get` and `app_timer_cnt_diff_compute` were supposed to be 'private' and should have been declared in the .c file instead of in the .h file...

Reply:
@Avamander can you explain why you add 1 in this line from StopWatch?
```cpp
delta += (currentTime + 1);
```

Reply:
Ok, changes are pushed. I'm using `TickType_t` internally for the timer controller but kept the inputs and outputs as ints so i don't have to include portmacro_cmsis.h in the app. I also did a quick sanity check to make sure that we cant have two overflows while the timer is running, which would make the remaining time ambiguous. The highest time that can be set are 59 minutes and 59 seconds, which is 3599 seconds in total. 1024*3599 is 0x383C00 which is less than (2^24-1) = 0xFFFFFF. So we're in the clear there. The time between overflows works out to be just about 4 hours and 33 minutes, so testing the overflow handling isn't really feasible. Best we can do is make sure it looks good ~~and wait for bug reports~~.

Reply:
Awesome, thanks!